src/executor/sort_merge_join.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+
+use std::cmp::Ordering;
+use std::vec::Vec;
+
+use futures::TryStreamExt;
+
+use super::*;
+use crate::array::{DataChunk, DataChunkBuilder, RowRef};
+use crate::types::DataType;
+
+/// The executor for sort merge join
+pub struct SortMergeJoinExecutor {
+    pub left_child: BoxedExecutor,
+    pub right_child: BoxedExecutor,
+    pub left_column_indexes: Vec<usize>,
+    pub right_column_indexes: Vec<usize>,
+    pub left_types: Vec<DataType>,
+    pub right_types: Vec<DataType>,
+}
+#[allow(dead_code)]
+impl SortMergeJoinExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
+    #[allow(dead_code)]
+    pub async fn execute(self) {
+        // collect all chunks from children
+        let (left_chunks, right_chunks) = async {
+            tokio::try_join!(
+                self.left_child.try_collect::<Vec<DataChunk>>(),
+                self.right_child.try_collect::<Vec<DataChunk>>(),
+            )
+        }
+        .await?;
+
+        // get rows iterator from chunks
+        let left_rows = || left_chunks.iter().flat_map(|chunk| chunk.rows()).peekable();
```

Comment:
What is the purpose of this closure? If I remove this closure and return the iterator trait directly .
    #[try_stream(boxed, ok = DataChunk, error = ExecutorError)] will throw this error 
`risinglight::executor
pub enum ExecutorError
The error type of execution.

[7 implementations]()

mismatched types
expected opaque type `impl Iterator<Item = data_chunk::RowRef<'_>>`
   found opaque type `impl Iterator<Item = data_chunk::RowRef<'_>>`rustc[E0308](https://doc.rust-lang.org/error-index.html#E0308)
[data_chunk.rs(60, 27): ]()the expected opaque type
[data_chunk.rs(60, 27): ]()the found opaque type
implementation of `FnOnce` is not general enough
closure with signature `fn(&'0 data_chunk::DataChunk) -> impl Iterator<Item = data_chunk::RowRef<'_>>` must implement `FnOnce<(&data_chunk::DataChunk,)>`, for any lifetime `'0`...
...but it actually implements `FnOnce<(&data_chunk::DataChunk,)>`

Reply:
The closure is used to flatten data chunk into rows with minimal overhead.

`left_chunks` is `Vec<DataChunk>`. We convert an iterator on data chunk to an iterator of `RowRef<'_>` using this function. There's no data clone within the process.

And your problem looks like a lifetime problem. Not sure what have you done, but it seems that the function signature is only valid for a single lifetime, instead of for all lifetimes.

By the way, if possible, I'd also prefer not to fetch all rows in advance. SortMergeJoin should be able to go over the data chunks from the child executors one by one.

Reply:
> By the way, if possible, I'd also prefer not to fetch all rows in advance. SortMergeJoin should be able to go over the data chunks from the child executors one by one.
I understand what you mean, loading all the chunks at once is not that necessary, I'll optimize it again